File: elixir/perfect-numbers/perfect_numbers.exs
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: ({ :ok, atom } | { :error, String.t() })
  def classify(number) when number <= 0 do
    {:error, "Classification is only possible for natural numbers." }
  end
  def classify(number) do
    factors_sum =
      number
      |> get_factors
      |> Enum.sum

    cond do
      factors_sum == number -> {:ok, :perfect}
      factors_sum < number -> {:ok, :deficient}
      factors_sum > number -> {:ok, :abundant}
    end
  end

  @spec get_factors(integer) :: [integer]
  def get_factors(1), do: [0]
  def get_factors(number) do
    for n <- (number-1)..1, rem(number, n) == 0, do: n
  end
end

File: python/perfect-numbers/perfect_numbers.py
def classify(number):
    if number < 1:
        raise ValueError('Must be more 1 or more.')

    aliquot_sum = sum(divisors(number)) - number
    if aliquot_sum == number:
        return 'perfect'
    elif aliquot_sum > number:
        return 'abundant'
    else:
        return 'deficient'


def divisors(n):
    # get factors and their counts
    factors = {}
    nn = n
    i = 2
    while i*i <= nn:
        while nn % i == 0:
            if i not in factors:
                factors[i] = 0
            factors[i] += 1
            nn //= i
        i += 1
    if nn > 1:
        factors[nn] = 1

    primes = list(factors.keys())

    # generates factors from primes[k:] subset
    def generate(k):
        if k == len(primes):
            yield 1
        else:
            rest = generate(k+1)
            prime = primes[k]
            for factor in rest:
                prime_to_i = 1
                # prime_to_i iterates prime**i values, i being all possible exponents
                for _ in range(factors[prime] + 1):
                    yield factor * prime_to_i
                    prime_to_i *= prime

    yield from generate(0)
